course_UvA-DL/transformers-and-MH-attention/Transformers_and_MHAttention.py

```diff
@@ -0,0 +1,1491 @@
+#
+# Below, we import our standard libraries.
+
+import json
+import math
+import os
+import random
+import urllib.request
+from functools import partial
+# %%
+# Standard libraries
+from urllib.error import HTTPError
+
+import matplotlib
+# Imports for plotting
+import matplotlib.pyplot as plt
+import numpy as np
+import pytorch_lightning as pl
+import seaborn as sns
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torch.utils.data as data
+import torchvision
+from IPython.display import set_matplotlib_formats
+from matplotlib.colors import to_rgb
+from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint
+from torchvision import transforms
+from torchvision.datasets import CIFAR100
+from tqdm.notebook import tqdm
+
+plt.set_cmap('cividis')
+# %matplotlib inline
+set_matplotlib_formats('svg', 'pdf')  # For export
+matplotlib.rcParams['lines.linewidth'] = 2.0
+sns.reset_orig()
+
+# tqdm for loading bars
+
+# PyTorch
+
+# Torchvision
+
+# PyTorch Lightning
+
+# Path to the folder where the datasets are/should be downloaded (e.g. CIFAR10)
+DATASET_PATH = os.environ.get('PATH_DATASETS', "data/")
+# Path to the folder where the pretrained models are saved
+CHECKPOINT_PATH = os.environ.get('PATH_CHECKPOINT', "saved_models/Transformers/")
+
+# Setting the seed
+pl.seed_everything(42)
+
+# Ensure that all operations are deterministic on GPU (if used) for reproducibility
+torch.backends.cudnn.determinstic = True
+torch.backends.cudnn.benchmark = False
+
+device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
+print("Device:", device)
+
+# %% [markdown]
+# Two pre-trained models are downloaded below.
+# Make sure to have adjusted your `CHECKPOINT_PATH` before running this code if not already done.
+
+# %%
+# Github URL where saved models are stored for this tutorial
+base_url = "https://raw.githubusercontent.com/phlippe/saved_models/main/tutorial6/"
+# Files to download
+pretrained_files = ["ReverseTask.ckpt", "SetAnomalyTask.ckpt"]
+
+# Create checkpoint path if it doesn't exist yet
+os.makedirs(CHECKPOINT_PATH, exist_ok=True)
+
+# For each file, check whether it already exists. If not, try downloading it.
+for file_name in pretrained_files:
+    file_path = os.path.join(CHECKPOINT_PATH, file_name)
+    if "/" in file_name:
+        os.makedirs(file_path.rsplit("/", 1)[0], exist_ok=True)
+    if not os.path.isfile(file_path):
+        file_url = base_url + file_name
+        print("Downloading %s..." % file_url)
+        try:
+            urllib.request.urlretrieve(file_url, file_path)
+        except HTTPError as e:
+            print(
+                "Something went wrong. Please try to download the file from the GDrive folder, or contact the author with the full output including the following error:\n",
```

Comment:
Why does it say gdrive when the models are on GH?

Reply:
I have the models as back-up as well on a gdrive, but we can also remove that out of the error message. I have not observed any issues with GH downloads before.